### DISCUSSION TITLE
MagickImage to Image control source

### DISCUSSION THREAD

**Discussion by nevaran** [dlemstra/Magick.NET#784] [General]
Im trying to figure out whats the most efficient/performant way to **view** the MagickImages but I cant seem to find any information on converting it so they can be assigned to an `Image.Source` .Is there some page that I am missing? If not, is there even a way?

**Comment by dlemstra** [comment]
I am guessing that you are using a WPF application so I assume that `Image.Source` is a `BitmapSource`. If you want to change the image into a `BitmapSource` you will need to install this library: https://www.nuget.org/packages/Magick.NET.SystemWindowsMedia/. And then you can do this `image.ToBitmapSource()`.

**Reply by nevaran** [reply]
It is BitmapImage, and yes, that seems to work; But it seems to have some memory leaking? Loading up a new image wont always release the previous one's memory and if its a large image, the app starts to use in the gigs (not to mention a bit too slow, even when resizing to screen resolution). I am using `.Freeze();` when I do ToBitmapSource, by the way.

And on a side note: Why does it seem to be using twice as much memory than doing it like this instead:
```
ToBitmapImage(image.ToByteArray())
public static BitmapImage ToBitmapImage(byte[] array)
        {
            using (var ms = new MemoryStream(array))
            {
                var image = new BitmapImage();
                image.BeginInit();
                image.CacheOption = BitmapCacheOption.OnLoad;
                image.StreamSource = ms;
                image.EndInit();
                image.Freeze();
                return image;
            }
        }
```

**Comment by dlemstra** [comment]
The BitMapSource that I create is not frozen: https://github.com/dlemstra/Magick.NET/blob/de73a5d374a3a78ca79f0b5f5fe78ecaef98ca03/src/Magick.NET.SystemWindowsMedia/IMagickImageExtentions.cs#L82. What happens when you `.Freeze()` it yourself?

It might be using more memory because the pixels are not compressed but raw? The `ToByteArray()` method will encode the image. Not sure what happens inside the `BitmapImage` that makes this use less memory though? Maybe the image is draw on the screen while the image is being decoded?

**Reply by nevaran** [reply]
When I freeze it i can `CG.Collect()` which helps somewhat when switching images; It also helps to be able to use async loading for thumbnail images.
I dont know if `ToByteArray` encodes the image, but it does take about twice as more time to load though, while taking half the memory after I return my `BitmapImage` which somewhat gives me the impression that `BitmapImage ` uses less resources than `BitmapSource`?

I dont know how Windows Photo Viewer manages to keep the memory usage insanely low (under 100mb) while loading 20,000x20,000 images, and while I would love to have such a performant app, I would prefer to load the actual image faster than use less memory (if possible?)

**Reply by dlemstra** [reply]
Maybe Windows Photo Viewer uses GPU memory instead?

p.s. I just stated that `ToByteArray` encodes the image.

**Reply by nevaran** [reply]
What does encoding it do , though? Does it have any specific benefits?

Also just tested it again, I noticed it goes from 6.2GB used memory to 7.5, it just apparently doesnt show on the Photo Viewer somehow
![image](https://user-images.githubusercontent.com/911128/99248077-95a2e300-2810-11eb-90e0-9ffa812aa853.png)

**Comment by nevaran** [comment]
I dont want to clutter the discussions with more posts so ill just ask it here: Can magick load gifs into a type of a control without needing to make it manually play the decoded frames one by one?

**Reply by dlemstra** [reply]
I don't know about a control that can do that. You can probably get some inspiration from this project: https://github.com/d2phap/ImageGlass.

**Reply by nevaran** [reply]
Alright, thanks
